### PR TITLE
[4063][ADD] stock_picking_location_check

### PR DESCRIPTION
**Pull request by AungKoKoLin1997**
[4063](https://www.quartile.co/web#id=4063&cids=3&menu_id=505&action=1457&model=project.task&view_type=form)

### CODE REVIEW

**Comment by AungKoKoLin1997** [comment]
We have to handle subcontracting process because the source location of receipt picking type and the source location of subcontracting move line are always different.

File: stock_picking_location_check/models/stock_picking.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    def _check_location_consistency(
+        self, pick_location, pick_dest_location, line_locations, line_dest_locations
+    ):
+        self.ensure_one()
+        if not (
+            set(line_locations).issubset(self._get_child_location_ids(pick_location))
+            and set(line_dest_locations).issubset(
+                self._get_child_location_ids(pick_dest_location)
+            )
+        ):
```
Comment (yostashiro) [inline stock_picking_location_check/models/stock_picking.py]: I suppose `_get_child_location_ids()` doesn't work recursively.

```suggestion
    def _check_location_consistency(self, pick_location, line_locations):
        self.ensure_one()
        if not (
            set(line_locations).issubset(
                pick_location.child_internal_location_ids or pick_location
            )
        ):
```

File: stock_picking_location_check/models/stock_picking.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    def _action_done(self):
+        for pick in self:
+            if pick.allow_location_inconsistency:
+                continue
+            self._check_location_consistency(
+                pick.location_id,
+                pick.location_dest_id,
+                pick.move_line_ids.location_id,
+                pick.move_line_ids.location_dest_id,
+            )
```
Comment (yostashiro) [inline stock_picking_location_check/models/stock_picking.py]:
```suggestion
            self._check_location_consistency(pick.location_id, pick.move_line_ids.location_id)
            self._check_location_consistency(pick.location_dest_id, pick.move_line_ids.location_dest_id)
```

File: stock_picking_location_check/models/stock_picking.py (hunk 1 above)
Comment (AungKoKoLin1997) [reply stock_picking_location_check/models/stock_picking.py]: `child_of` works recursively. I checked with hierarchical child locations and works well.
Reference: https://stackoverflow.com/questions/59562532/explanation-of-odoo-filter-domain-operator-child-of-and-its-preference-over-i

File: stock_picking_location_check/tests/test_stock_picking_location_check.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    def test_incoming_location_discrepancy_not_allowed(self):
+        self.create_move_and_move_line(
+            self.picking_in,
+            self.product,
+            self.supplier_location.id,
+            self.other_location.id,
+            10,
+        )
+        with self.assertRaises(UserError):
+            self.picking_in.button_validate()
```
Comment (yostashiro) [inline stock_picking_location_check/tests/test_stock_picking_location_check.py]: We should combine these two test.

```suggestion
    def test_incoming_location_discrepancy_check(self):
        self.create_move_and_move_line(
            self.picking_in,
            self.product,
            self.supplier_location.id,
            self.other_location.id,
            10,
        )
        with self.assertRaises(UserError):
            self.picking_in.button_validate()
        self.picking_in.allow_location_inconsistency = True
        self.picking_in.button_validate()
```

There is also not much point separating tests for incoming and outgoing, IMO. Shall we consolidate them as well?

File: stock_picking_location_check/tests/test_stock_picking_location_check.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.product = cls.env["product.product"].create({"name": "Test Product"})
+        cls.stock_location = cls.env.ref("stock.stock_location_stock")
+        cls.shelf1_location = cls.env.ref("stock.stock_location_components")
+        cls.shelf2_location = cls.env.ref("stock.stock_location_14")
+        cls.refrigerator_small_location = cls.env.ref(
+            "stock.location_refrigerator_small"
+        )
```
Comment (yostashiro) [inline stock_picking_location_check/tests/test_stock_picking_location_check.py]: We should avoid using the demo data: https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#avoid-relying-on-demo-data

The current test setup makes the assumption obscure (how can one tell that `cls.refrigerator_small_location` is a child of `cls.shelf2_location` without looking at the demo data definition?) and therefore is not desirable.

File: stock_picking_location_check/tests/test_stock_picking_location_check.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    def test_location_discrepancy(self):
+        # Test with discrepancy source location
+        picking_internal = self.create_picking(
+            self.other_location.id, self.shelf2_location.id
+        )
+        with self.assertRaises(UserError):
+            picking_internal.button_validate()
+        picking_internal.allow_location_inconsistency = True
+        picking_internal.button_validate()
+        # Test with discrepancy destination location
```
Comment (yostashiro) [inline stock_picking_location_check/tests/test_stock_picking_location_check.py]: We should split the test from here. Why combine?

**Review by yostashiro — CHANGES_REQUESTED** [review]

**Review by yostashiro — APPROVED** [review]

**Pull request merged**